### ISSUE TITLE
FFTA "link broken" when initiating any type of link

### ISSUE THREAD

**Issue by aawilson** [visualboyadvance-m/visualboyadvance-m#294]
I wanted to ask this on the forums, or at least search for the question there, but they are 404ing. I don't know how recent that is, or if it is expected.

### Expected behavior

When I enter the "trade items" menu from the link menu in Final Fantasy Tactics: Advance, it should initiate trade with the other, linked instance of VBA-M

### Actual behavior

FFTA generates the dialog, "Link broken, returning to world map"

### Steps to reproduce the behavior

0. Beforehand: VBA-M (v2.1.0) stock settings, except "Pause when inactive" in "Emulation" menu is deactivated, link type is set to "wireless", and "link timeout" in "options->link->configure" is set to "200"
1. Open two instances of VBA-M
2. Open FFTA rom (this was with EU version)
3. Import a working save file if it isn't already loaded (reset ROM after import)
4. Load the save in both instances
5. Enter the link menu from the world map
6. Initiate the link from "Options->Link->Start Network Link..." menu, with  one instance as client and the other as server. Observe successful connection.
7. Select "trade items" from the in-game link menu in both instances
8. The dialog above will be produced, and FFTA will exit to the world map

The above happens in both "wireless" and "cable" link mode.

### ROM details

Final Fantasy Tactics: Advance (Europe) (En, Fr, De, Es, It)

### Option details

Defaults, except as mentioned in first repro step above

### Build details

OS: Windows 10, 64-bit
Version of vba-m: 2.1.0

**Comment by aawilson** [comment]
Yeah, if you'll note, I tried both cable and wireless. But thanks for that and the other suggestion.

**Comment by aawilson** [comment]
Follow up question, you say "it works in linux". Do you specifically mean, linking with FFTA works in linux, or do you just mean linking works in linux?

**Comment by retro-wertz** [comment]
i havent tried with link with ffta. but i tested with four swords (vbam+dolphon) and aria or sorrows trade(gba+gba), so basic link should work

**Comment by aawilson** [comment]
There was some chatter in places when I searched around for this earlier (including a mysterious build including "code from VBAlink and VBA-m" that is no longer hosted anywhere) that seemed to indicate that FFTA was something of a special snowflake, any chance I could get a specific confirm for that ROM before I go digging around for something I can put linux and a window manager on?

**Comment by rkitover** [comment]
forums indicate that FFTA should work.

@denisfa and I were going to investigate link on windows further now that we have significantly improved windows development environment support.

**Comment by denisfa** [comment]
@rkitover I can confirm this is working on `Windows 10` using `Cable` and timeout `200` with the latest [nightly](https://github.com/visualboyadvance-m/visualboyadvance-m/issues/354#issuecomment-568932350). No special configuration required.

![ffta2](https://user-images.githubusercontent.com/11354751/72743696-e6af5f80-3b8a-11ea-90af-df3470a49919.PNG)

Using `Wireless`, the player 2 gets lagged to oblivion. I am getting 2 FPS on average and being spammed by log messages about the link. I cannot do anything, but also did not get the error message from the game.

Using the following sav file for both connections (removing the `.txt` extension):
[final_fantasy_tactics_advance-1.sav.txt](https://github.com/visualboyadvance-m/visualboyadvance-m/files/4087135/final_fantasy_tactics_advance-1.sav.txt)

**Comment by rkitover** [comment]
@denisfa thanks for testing, did you by any chance try the specific error condition, which is trading items in the game between the two instances?

I can give that a try, I wanted to start working on the link code anyway, because fixing the link bugs are about the most wanted feature in this project.

**Comment by denisfa** [comment]
@rkitover Yes, I did. I can add the screenshots if necessary. On both Windows and Linux with the above requisites.

**Comment by ZachBacon** [comment]
Might be a good thing to have in the wiki once we get it going.

**Comment by rkitover** [comment]
@denisfa awesome, I don't even know which specific link things we need to fix, I just no of the menu pause thing, which I was going to look at at some point, but that one is not going to be easy.

**Comment by denisfa** [comment]
@rkitover One thing we **REALLY NEED** is to be able to reset the connection. Something like when you set to `nothing` it closes and cleans everything from the current connection. For both server and client. It is so annoying to close and open the emulator every time we want to check something.

**Comment by rkitover** [comment]
@denisfa gotcha, I'll add that to my list.

**Comment by JackoboLeChocobo** [comment]
If FFTA hangs in link mode, the main issue is ( or was if the addon was added in this project ) the General Purpose Mode that was not added.

Source : https://problemkaputt.de/gbatek.htm

```
SIO General-Purpose Mode

In this mode, the SIO is 'misused' as a 4bit bi-directional parallel port, each of the SI,SO,SC,SD pins may be directly controlled, each can be separately declared as input (with internal pull-up) or as output signal.

4000134h - RCNT (R) - SIO Mode, usage in GENERAL-PURPOSE Mode (R/W)
Interrupts can be requested when SI changes from HIGH to LOW, as General Purpose mode does not require a serial shift clock, this interrupt may be produced even when the GBA is in Stop (low power standby) state.

  Bit   Expl.
  0     SC Data Bit         (0=Low, 1=High)
  1     SD Data Bit         (0=Low, 1=High)
  2     SI Data Bit         (0=Low, 1=High)
  3     SO Data Bit         (0=Low, 1=High)
  4     SC Direction        (0=Input, 1=Output)
  5     SD Direction        (0=Input, 1=Output)
  6     SI Direction        (0=Input, 1=Output, but see below)
  7     SO Direction        (0=Input, 1=Output)
  8     SI Interrupt Enable (0=Disable, 1=Enable)
  9-13  Not used
  14    Must be "0" for General-Purpose Mode
  15    Must be "1" for General-Purpose or JOYBUS Mode

SI should be always used as Input to avoid problems with other hardware which does not expect data to be output there.

```

In my project, I have solved this issue with this code on GBALink.cpp added in void ```
StartGPLink(u16 value) : 
case GP:
		if(linklog){ 
			if(value==0x8000) fprintf(jjj, "Circuit reset\n");
			else /*if(!rfu_enabled)*/ fprintf(jjj, "Using General-purpose mode %04x\n", value);
		}
		if(value==0x8000) UPDATE_REG(0x134, value); //Circuit Reset
		//else if (rfu_enabled) rfu_state = RFU_INIT;
		else //General Purpose Mode
			{
			UPDATE_REG(0x134, value);
			if(value==0x8055) value=0x8052;
			else if(value==0x8022) value=0x8025;
			else if(value==0x8050) value=0x8050;	
			else if(value==0x8020) value=0x8020;
			UPDATE_REG(0x134, value);
			}
		break;
```

**Comment by rkitover** [comment]
@JackoboLeChocobo thank you very much for the info, I'll try to add this ASAP.

**Comment by JackoboLeChocobo** [comment]
@rkitover No problem ^^ ( the values 0x8055 0x8052 0x8022 0x8025 0x8050 and 0x8020 are the values used by FFTA. In my knownledge, it is the only game that I see using the General Purpose Mode )

**Comment by RaythaLaws95** [comment]
Hi, Devs

This my first time using github and I'm no programmer so hopefully I'm not breaking any rules here.

I was playing FFTA (US, Australian ver) on vba-m (2.1.4) and I tried the link feature. I followed the settings you discussed above, it worked at the beginning but the issue I encountered was, for every link mission, as soon as any of my character perform any action (normal attack, A-skill), the link breaks and I get the error message of "link broken". May I know if there is any way to fix it?

Thank you in advance!

**Comment by Reave881** [comment]
Same issues as @RaythaLaws95

Trading items and trading units works well,

But any teaming up mission leads to a crash upon the use of any abilities in the missions.

From my readings left and right it was a FFTA-specific issue, but I couldn't find any solution and VBA-Link says on their page that it flat out does not support FFTA/General Purpose Mode.

**Comment by RoxNdroid** [comment]
Same issue as @RaythaLaws95 . 
and I gotta add that the disconnect occurs whenever the enemy is damaged as well, even if it’s after a counter.

Have any of you found a way around it? I’m still trying to figure out how to apply @ghost ‘s fix

**Comment by leoxxx** [comment]
No$GBA can link the game.
So as per NO$GBA.

**Comment by Squall-Leonhart** [comment]
we have a branch with this fixed, but i'm not sure what reasons for it not being merged into master are about.

**Comment by jleenerts** [comment]
I just tried building the latest master branch with the replaced code from the ffta-link branch ([this commit](https://github.com/visualboyadvance-m/visualboyadvance-m/commit/f015f743b792574b410119dc70c746e3159ca607)) and I can confirm that it works. Hopefully it will get merged so more people can try out the multiplayer. Without it, I was only able to trade units and items (and even then, it would sometimes take significant time to verify and complete the transaction).

**Comment by Squall-Leonhart** [comment]
it won't be merged, its a hack.

**Comment by jleenerts** [comment]
Ah, I apologize for the misunderstanding. From your last comment it sounded like it was intended to be merged as a fix.

**Comment by andyvand** [comment]
It should be fixed and needs testing. Latest nightly should work. it has GP mode support and IPC/LAN link fixes
